Add error-specific validation message in site identification page

Imran92 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/91479

## Proposed Changes

Added error-specific validation messages for the following scenarios.

1. Input is an Email
2. Input has invalid characters
3. Wrong Protocol (FTP://, file:// etc.)
4. Missing TLD
5. IDN URLs (Either has non ASCII characters in URL or hostname starts with xn--)
6. General error message for fallback

I have replaced "Incomplete URL Error" from the issue with Wrong Protocol because otherwise the error is too ambiguous and broad to detect. But I am a bit unsure about using the word "Protocol" in the error message, looks too technical. I only kept it because we show it only when the user has already used the wrong protocol and it conveys the exact reason. But LMK if you think a different and less technical message should be shown.

Also, detecting missing TLD is super tricky. For example, "test.com" is valid but "test.wpcomstaging" is not even though both of them has the same structure as a string. So it's not too correct when implemented with regex or some other way through code. But it should be helpful because our error log says users mostly write URLs like this "myblog" instead of "myblog.com" in which case, it'll be helpful.

Also, I've included an example of a valid URL with each error message to better convey to the user how a URL looks.

## Why are these changes being made?
To enable the user to better understand what's wrong with the URL they have input.

## Screenshots

![Screenshot 2024-06-18 at 7 58 38 AM](https://github.com/Automattic/wp-calypso/assets/1190420/f5011e43-b26b-47ee-8583-92f4ee430eb3)

![Screenshot 2024-06-18 at 7 59 14 AM](https://github.com/Automattic/wp-calypso/assets/1190420/89ac6da2-365d-4c79-8edc-a630d0645be3)

![Screenshot 2024-06-18 at 7 59 38 AM](https://github.com/Automattic/wp-calypso/assets/1190420/1388b421-64d0-47df-ab25-e338b9a09442)

![Screenshot 2024-06-18 at 8 00 05 AM](https://github.com/Automattic/wp-calypso/assets/1190420/52d4aadf-9d5e-492c-a0fc-dc4f6dfb9fc9)

![Screenshot 2024-06-18 at 8 00 32 AM](https://github.com/Automattic/wp-calypso/assets/1190420/e0612f33-b7a3-45b3-b21c-faf246908fd4)

![Screenshot 2024-06-18 at 8 01 10 AM](https://github.com/Automattic/wp-calypso/assets/1190420/7a0179c7-31a5-4c12-9e19-8d3ed34accfe)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Make sure your language is English in https://wordpress.com/me/account
- In the migration flow (take any migration path, for example http://calypso.localhost:3000/start), navigate to the Site Identification step
- In the URL field, try different wrong URLs (Please check the **Proposed Changes** section above). For example - myblog, https://myblog, https://myblog^, email@myblog.com etc.
- You should see different error messages depending on the URL
- Change the language to some other language in `https://wordpress.com/me/account`
- Go to the site identification page and try different wrong URLs again
- Make sure you see the old error message

https://github.com/Automattic/wp-calypso/assets/6820724/4d46c197-2ac1-4ff6-98e7-d80b5034da9c

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

Imran92 committed:
Show dynamic validation message for URL

Imran92 committed:
Should show the TLD error when applicable

Imran92 committed:
Should show TLD error when protocol is given

Imran92 committed:
Add test for all error messages

Imran92 committed:
Update test URL for default case

Imran92 committed:
Fix type for url

donnapep commented:
WDYT about running these error message through content review first pfGA22-2-p2 ?

Imran92 committed:
Fix e2e error

Imran92 committed:
Add period at the end of the copies

Imran92 commented:
Thanks @donnapep! All tests are green now.

> WDYT about requesting a copy review (pfGA22-2-p2) of these error messages first?

This is a TIL, thanks for sharing it! Just wondering if it would be worth running these through Bohdan/Jana first before submitting for review, WDYT?

donnapep commented:
> This is a TIL, thanks for sharing it! Just wondering if it would be worth running these through Bohdan/Jana first before submitting for review, WDYT?

We can just go straight to the copy review folks as they're the copy experts. 🙂

Imran92 commented:
> We can just go straight to the copy review folks as they're the copy experts. 🙂

Thanks, on it!

donnapep commented:
👋🏻 I did a quick test of the behavior. I found it somewhat distracting to see the error message update as I typed. Could we change it so that the message only changes after clicking the button?

Imran92 committed:
Show error message only after submission

Imran92 commented:
> 👋🏻 I did a quick test of the behavior. I found it somewhat distracting to see the error message update as I typed. Could we change it so that the message only changes after clicking the button?

Thanks for checking @donnapep, I've changed the behavior here https://github.com/Automattic/wp-calypso/pull/91748/commits/f5b85cfb53823048f2affbf852ad8be2d7d13953 so that it only appears when you click the button and message disappears when you change something in the input field. I've seen this behavior on many login pages. In case you prefer this message to stay until the next click, LMK!

Imran92 committed:
Detect IDN and show proper validation message

Imran92 committed:
Add test for IDN urls

Imran92 committed:
Fix url type

gabrielcaires commented on client/blocks/import/capture/capture-input.tsx:
```diff
@@ -69,13 +72,124 @@ const CaptureInput: FunctionComponent< Props > = ( props ) => {
 		}
 	}
 
+	function isIDN( url: string ) {
```
Can we have this function outside the component body?
Each render is recreating the function.

Imran92 replied:
Thanks, I've handled it here https://github.com/Automattic/wp-calypso/pull/91748/commits/228d668523ae30f2e7df7fbf5e4a90c0518c4348

gabrielcaires commented on client/blocks/import/capture/capture-input.tsx:
```diff
@@ -69,13 +72,124 @@ const CaptureInput: FunctionComponent< Props > = ( props ) => {
+	}
+
+	// TODO: Just return the translated string directly from getValidationMessage once we have translations for all messages.
+	const errorMessages = {
```
Suggestion: Move all error message logic to a private component, since is this pretty big.

Imran92 replied:
Thank you, done here https://github.com/Automattic/wp-calypso/pull/91748/commits/228d668523ae30f2e7df7fbf5e4a90c0518c4348

gabrielcaires commented on client/blocks/import/capture/test/index.tsx:
```diff
@@ -85,4 +85,149 @@ describe( 'CaptureInput', () => {
+		);
+	} );
+
+	test( 'should show error for missing top-level domain', async () => {
```
Usually, we use the `it` instead of `test` the file already has scenarios with `it`. 

```suggestion
	it.( 'shows error for missing top-level domain', async () => {
```

Imran92 replied:
Done under it.each here https://github.com/Automattic/wp-calypso/pull/91748/commits/3447cd32549e85efbeeba51ca3f8ec7882a64e3d

gabrielcaires commented on client/blocks/import/capture/test/index.tsx:
```diff
@@ -85,4 +85,149 @@ describe( 'CaptureInput', () => {
+	} );
+
+	test( 'should show error for missing top-level domain event when protocol is present', async () => {
+		await enterUrlAndContinue( 'https://myblog' );
```
You can use dinamic to test all scenarios:


Example:
```
it.each([
  {scenario: 'A', input: 'XXX',  error: 'YYY'},
  {scenario: 'B', input: 'BB',  error: 'ZZZ'},
])(' show error ', ({scenario,  input, expected}) => {
    await enterUrlAndContinue( input );
expect(
			screen.getByText( error )
		).toBeInTheDocument();
});
```

Docs here https://jestjs.io/docs/api#testeachtablename-fn-timeout

Imran92 replied:
Very good suggestion, thank you! Done here https://github.com/Automattic/wp-calypso/pull/91748/commits/3447cd32549e85efbeeba51ca3f8ec7882a64e3d

gabrielcaires reviewed (changes requested):
I left some suggestions

Imran92 committed:
Move validation logic to separate file

Imran92 committed:
Change tests to use it.each

Imran92 committed:
Update type to fix error

Imran92 committed:
Use updated validation messages

donnapep committed:
Update IDN error message

donnapep committed:
Shorten invalid protocol error message and fix quotation marks

donnapep commented:
Updated error messages in PR description based on 1NK0CW5QEL336zXL4QfQdiNCDNQhVO2uWJbjEap3rLNM-gdoc.

donnapep commented:
E2E errors are not related to this PR and are happening for all PRs.

donnapep commented:
> Also, detecting missing TLD is super tricky. For example, "test.com" is valid but "test.wpcomstaging" is not even though both of them has the same structure as a string. 

Is there any chance we would show the message even if the URL was valid, and so the user wouldn't be able to move on? Because if so, that would be bad. 🙅🏻 

But I'm thinking that it would be more the opposite case where we wouldn't catch some missing TLDs, so the user would see the default error message instead. Does that all sound correct?

donnapep commented on client/blocks/import/capture/url-validation-message-helper.ts:
```diff
@@ -0,0 +1,123 @@
+			message: translate(
+				"Looks like your site address is missing its domain extension. Please try again with something like 'example.com' or 'example.net'."
+			),
+			messageString:
```
Are we duplicating the strings here only for translation purposes? I think so, and I'm kinda tempted to add the "String Freeze" label while you're away.

donnapep commented on client/blocks/import/capture/url-validation-message-helper.ts:
```diff
@@ -0,0 +1,123 @@
+	hasEnTranslation: ( key: string ) => boolean
+) {
+	const errorMessages = getErrorMessages( translate );
+	const hasEnTranslationForAllMessages = Object.values( errorMessages ).every( ( message ) =>
```
Curious as to why this is all-or-nothing? I would think that we would want to show an error message if it's been translated, even if the others have not.

donnapep commented:
I couldn't get the _Looks like your URL has some invalid characters (like ~ or ^)._ message to trigger with
 the following:

![Screenshot 2024-06-17 at 3 32 08 PM](https://github.com/Automattic/wp-calypso/assets/1190420/86d53d1b-109f-415c-9418-e9e652ce2ad3)

donnapep commented:
@Imran92 Given the strings have already been approved in 1NK0CW5QEL336zXL4QfQdiNCDNQhVO2uWJbjEap3rLNM-gdoc, I'm going ahead and adding the "String Freeze" label. Hopefully they'll be translated by the time you're back, and then we can get rid of the additional complexity / duplicated strings prior to merge.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/14781025

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Hi @Imran92, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include all of the following strings:
- `Looks like your site address is missing its domain extension. Please try again with something like 'example.com' or 'example.net'.`
- `Looks like you might have added an email address. Please use a URL instead, like 'example.com'.`
- `Looks like your URL has some invalid characters (like ~ or ^). Please delete them and try again.`
- `URLs usually start with http:// or https:// (rather than file:/, ftp://, or similar). Please try again with a URL like 'https://example.com'.`
- `Looks like you’ve added an internationalized domain name. Please try a standard URL instead (like 'example.com').`
- `Please add a valid website address (like 'example.com'). Feel free to copy and paste it in if that helps.`

Thank you in advance!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/14781025).

Imran92 commented:
> Is there any chance we would show the message even if the URL was valid, and so the user wouldn't be able to move on? Because if so, that would be bad. 🙅🏻

Very good and valid question, yep that'll be bad if that happens. Actually, this was my first fear as well, so to completely avoid that, I only try to detect the type of error when the existing main validation regex detects an error. Only when the `isError` flag is true, then we try to figure out the proper message for it -

```
const tempValidationMessage = isValid
			? ''
			: getValidationMessage( url, translate, hasEnTranslation );

```

> But I'm thinking that it would be more the opposite case where we wouldn't catch some missing TLDs, so the user would see the default error message instead. Does that all sound correct?

Yap, Right, that will probably be the case

Imran92 commented:
> Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/14781025).

I am seeing no associated translation strings here https://translate.wordpress.com/deliverables/overview/14781025 .

Imran92 committed:
Remove extra strings for translation check

We are using the string freeze label instead

donnapep commented:
I re-tested and can confirm that the internationalized domain name message and the generic error message are showing in English, so there's some problem here:

![Screenshot 2024-06-24 at 11 37 56 AM](https://github.com/Automattic/wp-calypso/assets/1190420/f3c7a2f2-9192-4caa-81fe-934f0d32c63d)

![Screenshot 2024-06-24 at 11 38 07 AM](https://github.com/Automattic/wp-calypso/assets/1190420/7a2107e8-ab88-498f-a563-7ac952930774)

Also, any thoughts on https://github.com/Automattic/wp-calypso/pull/91748#issuecomment-2174285008?

donnapep commented:
For French, none of the error messages are translated.

Imran92 committed:
Remove tilde character from accepted list

Imran92 commented:
> I couldn't get the _Looks like your URL has some invalid characters (like ~ or ^)._ message to trigger with the following:

Thanks for finding it Donna, fixed it [here](https://github.com/Automattic/wp-calypso/pull/91748/commits/aac5931c8d5aefdcb130d14a069cd7b4e582078f)

> Also, I had some strange behavior when testing that is not related to this PR, but perhaps you may still have insight into. I added `supsup.com` as the URL, and on the next screen I clicked the _Back_ button. When I went to the previous screen, my URL was now set to `https://www.brandbucket.com/names/supsup?source=ext/`, which is really weird. Wondering if we saw a URL like that at all in the error data? Seems like a potential bug to me: 

Nice finding! I wasn't sure what was happening here, so I investigated the issue and found the reason. Apparently, if the URL that is put in the input field redirects to some other URL, we replace the input URL with the final redirected URL. The URL you put there (supsup.com) is on sell by brandbucket.com, so if anyone tries to access the URL, they redirect it to the purchase page. The same will happen if you try with "dubico.com" or "yazani.com" etc. Same happens if we try to access a WPCOM site with the wpcomstaging URL when it has a different primary domain. I'm not sure if it is an expected behavior or a bug though :p

donnapep commented:
@Imran92 What's the plan for figuring out the translation issue? If we can't solve it in a timely manner, then let's revert the change that removed the use of `useHasEnTranslation` and ship for English users only. Either way, let's please ship by the end of the week. 🙇🏻‍♀️

donnapep reviewed (approved):
Everything looks good! Let's :shipit:
